### PR TITLE
YOLO Family: Darknet backbones

### PR DESCRIPTION
**Pull request by anivegesana**
# Description

This pull request adds the backbones of the YOLOv3 and v4 families (Darknet53, TinyYOLOv3, CSPDarknet53, and TinyYOLOv4).

To train the CSPDarknet53 image classification backbone on ImageNet, use the following command in the official directory:
```
python3 -m official.vision.beta.projects.yolo.train --mode=train_and_eval --experiment=darknet_classification --model_dir=training_dir --config_file=official/vision/beta/projects/yolo/configs/experiments/darknet53.yaml
```

## Type of change

For a new feature or function, please create an issue first to discuss it
with us before submitting a pull request.

Note: Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation update
- [ ] TensorFlow 2 migration
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] A new research paper code implementation
- [ ] Other (Specify)

## Tests

Tests are provided for each of the building blocks. The tests provide checking to see if the outputs or gradients of the layers go to `None` anywhere in the layer.

**Test Configuration**: Currently, we use `unittest` in the Python3 standard library in order to verify the correctness of our code.

## Checklist

- [x] I have signed the [Contributor License Agreement](https://github.com/tensorflow/models/wiki/Contributor-License-Agreements).
- [x] I have read [guidelines for pull request](https://github.com/tensorflow/models/wiki/Submitting-a-pull-request).
- [x] My code follows the [coding guidelines](https://github.com/tensorflow/models/wiki/Coding-guidelines).
- [x] I have performed a self [code review](https://github.com/tensorflow/models/wiki/Code-review) of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.

### CODE REVIEW

**Comment by vishnubanna** [comment]
@googlebot I signed it!

**Comment by The-Indian-Chinna** [comment]
@googlebot I signed it!

**Comment by anivegesana** [comment]
@googlebot I signed the CLA

**Comment by vishnubanna** [comment]
@googlebot I consent.

**Comment by anivegesana** [comment]
@googlebot I consent.

**Comment by The-Indian-Chinna** [comment]
@googlebot I consent.

File: official/vision/beta/projects/yolo/README.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# YOLO Object Detectors, You Only Look Once
+
+[![Paper](http://img.shields.io/badge/Paper-arXiv.1804.02767-B3181B?logo=arXiv)](https://arxiv.org/abs/1804.02767)
+[![Paper](http://img.shields.io/badge/Paper-arXiv.2004.10934-B3181B?logo=arXiv)](https://arxiv.org/abs/2004.10934)
+
+This repository is the unofficial implementation of the following paper. However, we spent painstaking hours ensuring that every aspect that we constructed was the exact same as the original paper and the original repository.
```
Comment (davisjam) [inline official/vision/beta/projects/yolo/README.md:6]: Second sentence seems unnecessary.

Comment (davisjam) [inline official/vision/beta/projects/yolo/README.md:6]: First sentence says "paper" but there are two.

File: official/vision/beta/projects/yolo/README.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+* YOLOv3: An Incremental Improvement: [YOLOv3: An Incremental Improvement](https://arxiv.org/abs/1804.02767)
```
Comment (davisjam) [inline official/vision/beta/projects/yolo/README.md:8]: Maybe this is garden convention, but not sure why you have both plaintext and then hypertext versions of title

File: official/vision/beta/projects/yolo/README.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+Yolo v3 and v4 serve as the most up to date and capable versions of the Yolo network group. These model uses a custom backbone called Darknet53 that uses knowledge gained from the ResNet paper to improve its predictions. The new backbone also allows for objects to be detected at multiple scales. As for the new detection head, the model now predicts the bounding boxes using a set of anchor box priors (Anchor Boxes) as suggestions. The multiscale predictions in combination with the Anchor boxes allows for the network to make up to 1000 object predictions on a single image. Finally, the new loss function forces the network to make better prediction by using Intersection Over Union (IOU) to inform the models confidence rather than relying on the mean squared error for the entire output.
```
Comment (davisjam) [inline official/vision/beta/projects/yolo/README.md:18]: "model's" or "models'" or "models"?

File: official/vision/beta/projects/yolo/README.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+## Table of Contents
+
+* [Our Goal](#our-goal)
+* [Models in the library](#models-in-the-library)
+* [References](#references)
+
+
+## Our Goal
+Our goal with this model conversion is to provide highly versatile implementations of the Backbone and Yolo Head. We have tried to build the model in such a way that the Yolo head could easily be connected to a new, more powerful backbone if a person chose to.
```
Comment (davisjam) [inline official/vision/beta/projects/yolo/README.md:36]: Inconsistent formatting, should there be a newline here?

Comment (davisjam) [inline official/vision/beta/projects/yolo/README.md:36]: I would remove the rhetorical flourishes like "highly versatile".

File: official/vision/beta/projects/yolo/common/__init__.py [hunk 5]
```diff
@@ -0,0 +1 @@
+
```
Comment (davisjam) [inline official/vision/beta/projects/yolo/common/__init__.py:1]: Is this a typo?

File: official/vision/beta/projects/yolo/configs/backbones.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
+"""Backbones configurations."""
+# Import libraries
+import dataclasses
+from typing import Optional
+from official.modeling import hyperparams
+
+# from official.vision.beta.configs import backbones
```
Comment (davisjam) [inline official/vision/beta/projects/yolo/configs/backbones.py:7]: Remove unused code

File: official/vision/beta/projects/yolo/configs/backbones.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+
+@dataclasses.dataclass
+class DarkNet(hyperparams.Config):
+  """DarkNet config."""
+  model_id: str = "darknet53"
+
+
+# # we could not get this to work
```
Comment (davisjam) [inline official/vision/beta/projects/yolo/configs/backbones.py:16]: Remove unused code

File: official/vision/beta/projects/yolo/configs/experiments/darknet53.yaml [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+      type: 'polynomial'
+      polynomial:
+        initial_learning_rate: 0.1
+        end_learning_rate: 0.0001
+        power: 4.0 
+        decay_steps: 51136000
+    warmup:
+      type: 'linear'
+      linear:
+        warmup_steps: 64000 #lr rise from 0 to 0.1 over 1000 steps
```
Comment (davisjam) [inline official/vision/beta/projects/yolo/configs/experiments/darknet53.yaml:53]: Not sure of the notation, but perhaps this should say not "#lr" but "# linear" ?

File: official/vision/beta/projects/yolo/modeling/__init__.py [hunk 9]
```diff
@@ -0,0 +1 @@
+#from .yolo_v3 import Yolov3
```
Comment (davisjam) [inline official/vision/beta/projects/yolo/modeling/__init__.py:1]: Is this change necessary?

File: official/vision/beta/projects/yolo/modeling/backbones/Darknet.py [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+                            config,
+                            name=f"{config.layer}_{i}")
+        stack_outputs.append(x)
+      elif config.stack == "csp_tiny":
+        x_pass, x = self._tiny_stack(stack_outputs[config.route],
+                                     config,
+                                     name=f"{config.layer}_{i}")
+        stack_outputs.append(x_pass)
+      if (config.is_output and
+          self._min_size == None):  # or isinstance(config.output_name, str):
```
Comment (davisjam) [inline official/vision/beta/projects/yolo/modeling/backbones/Darknet.py:253]: Is this comment needed?

File: official/vision/beta/projects/yolo/modeling/building_blocks/_CSPConnect.py [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+  def __init__(
+      self,
+      filters,
+      filter_reduce=2,
+      activation="mish",
+      kernel_initializer='glorot_uniform',
+      bias_initializer='zeros',
+      bias_regularizer=None,
+      weight_decay=None,  # default find where is it is stated
```
Comment (davisjam) [inline official/vision/beta/projects/yolo/modeling/building_blocks/_CSPConnect.py:17]: Comment grammar

File: official/vision/beta/projects/yolo/modeling/building_blocks/_CSPDownSample.py [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+  def __init__(
+      self,
+      filters,
+      filter_reduce=2,
+      activation="mish",
+      kernel_initializer='glorot_uniform',
+      bias_initializer='zeros',
+      bias_regularizer=None,
+      weight_decay=None,  # default find where is it is stated
```
Comment (davisjam) [inline official/vision/beta/projects/yolo/modeling/building_blocks/_CSPDownSample.py:17]: Comment grammar

File: official/vision/beta/projects/yolo/modeling/building_blocks/_DarkConv.py [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+      weight_decay=None,  # default find where is it is stated
+      use_bn=True,
+      use_sync_bn=False,
+      norm_momentum=0.99,
+      norm_epsilon=0.001,
+      activation='leaky',
+      leaky_alpha=0.1,
+      **kwargs):
+    '''
+        Modified Convolution layer to match that of the DarkNet Library
```
Comment (davisjam) [inline official/vision/beta/projects/yolo/modeling/building_blocks/_DarkConv.py:35]: Can you clarify a little more how this is different?

File: official/vision/beta/projects/yolo/modeling/building_blocks/_DarkConv.py [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+
+    super(DarkConv, self).__init__(**kwargs)
+    return
+
+  def build(self, input_shape):
+    kernel_size = self._kernel_size if type(
+        self._kernel_size) == int else self._kernel_size[0]
+    if self._padding == "same" and kernel_size != 1:
+      self._zeropad = ks.layers.ZeroPadding2D(
+          ((1, 1), (1, 1)))  # symetric padding
```
Comment (davisjam) [inline official/vision/beta/projects/yolo/modeling/building_blocks/_DarkConv.py:102]: symmetric

File: official/vision/beta/projects/yolo/modeling/building_blocks/_DarkConv.py [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+    else:
+      self._zeropad = Identity()
+
+    self.conv = ks.layers.Conv2D(
+        filters=self._filters,
+        kernel_size=self._kernel_size,
+        strides=self._strides,
+        padding="valid",  #self._padding,
```
Comment (davisjam) [inline official/vision/beta/projects/yolo/modeling/building_blocks/_DarkConv.py:110]: What is this comment for?

File: official/vision/beta/projects/yolo/modeling/building_blocks/_DarkConv.py [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+
+  def call(self, inputs):
+    x = self._zeropad(inputs)
+    x = self.conv(x)
+    x = self.bn(x)
+    x = self._activation_fn(x)
+    return x
+
+  def get_config(self):
+    # used to store/share parameters to reconsturct the model
```
Comment (davisjam) [inline official/vision/beta/projects/yolo/modeling/building_blocks/_DarkConv.py:151]: reconstruct

File: official/vision/beta/projects/yolo/modeling/building_blocks/_DarkResidual.py [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+
+  def call(self, inputs):
+    shortcut = self._dconv(inputs)
+    x = self._conv1(shortcut)
+    x = self._conv2(x)
+    x = self._shortcut([x, shortcut])
+    return self._activation_fn(x)
+
+  def get_config(self):
+    # used to store/share parameters to reconsturct the model
```
Comment (davisjam) [inline official/vision/beta/projects/yolo/modeling/building_blocks/_DarkResidual.py:141]: reconstruct

File: official/vision/beta/projects/yolo/modeling/tests/test_DarkConv.py [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+
+    with tf.GradientTape() as tape:
+      x_hat = test_layer(x)
+      grad_loss = loss(x_hat, y)
+    grad = tape.gradient(grad_loss, test_layer.trainable_variables)
+    optimizer.apply_gradients(zip(grad, test_layer.trainable_variables))
+    self.assertNotIn(None, grad)
+    return
+
+  # @parameterized.named_parameters(("filters", 3), ("filters", 20), ("filters", 512))
```
Comment (davisjam) [inline official/vision/beta/projects/yolo/modeling/tests/test_DarkConv.py:57]: What is this for?

Comment (vishnubanna) [reply official/vision/beta/projects/yolo/modeling/tests/test_DarkConv.py]: fixed

File: official/vision/beta/projects/yolo/modeling/building_blocks/_DarkResidual.py (hunk 17 above)
Comment (vishnubanna) [reply official/vision/beta/projects/yolo/modeling/building_blocks/_DarkResidual.py]: fixed

File: official/vision/beta/projects/yolo/modeling/building_blocks/_DarkConv.py (hunk 16 above)
Comment (vishnubanna) [reply official/vision/beta/projects/yolo/modeling/building_blocks/_DarkConv.py]: fixed

File: official/vision/beta/projects/yolo/modeling/backbones/Darknet.py [hunk 19]
```diff
@@ -0,0 +1,356 @@
...
+
+def darktiny_config_todict(config, kwargs):
+  dictvals = {"filters": config.filters, "strides": config.strides}
+  dictvals.update(kwargs)
+  return dictvals
+
+
+def maxpool_config_todict(config, kwargs):
+  return {
+      "pool_size": config.kernel_size,
```
Comment (arashwan) [inline official/vision/beta/projects/yolo/modeling/backbones/Darknet.py:66]: Uhmm, I feel this is a bit misuse of the naming, can you introduce pool_size to block config list?

File: official/vision/beta/projects/yolo/modeling/backbones/Darknet.py [hunk 20]
```diff
@@ -0,0 +1,356 @@
...
+
+class layer_registry(object):
+
+  def __init__(self):
+    self._layer_dict = {
+        "DarkTiny": (nn_blocks.DarkTiny, darktiny_config_todict),
+        "DarkConv": (nn_blocks.DarkConv, darkconv_config_todict),
+        "MaxPool": (tf.keras.layers.MaxPool2D, maxpool_config_todict)
+    }
+    return
```
Comment (arashwan) [inline official/vision/beta/projects/yolo/modeling/backbones/Darknet.py:81]: remove return?

File: official/vision/beta/projects/yolo/modeling/backbones/Darknet.py [hunk 21]
```diff
@@ -0,0 +1,356 @@
...
+  return dictvals
+
+
+def darktiny_config_todict(config, kwargs):
+  dictvals = {"filters": config.filters, "strides": config.strides}
+  dictvals.update(kwargs)
+  return dictvals
+
+
+def maxpool_config_todict(config, kwargs):
```
Comment (arashwan) [inline official/vision/beta/projects/yolo/modeling/backbones/Darknet.py:64]: Can you add darktiny_config_todict and maxpool_config_todict as member functions for layer_registry?

File: official/vision/beta/projects/yolo/modeling/backbones/Darknet.py [hunk 22]
```diff
@@ -0,0 +1,356 @@
...
+
+  def __init__(self):
+    self._layer_dict = {
+        "DarkTiny": (nn_blocks.DarkTiny, darktiny_config_todict),
+        "DarkConv": (nn_blocks.DarkConv, darkconv_config_todict),
+        "MaxPool": (tf.keras.layers.MaxPool2D, maxpool_config_todict)
+    }
+    return
+
+  def _get_layer(self, key):
```
Comment (arashwan) [inline official/vision/beta/projects/yolo/modeling/backbones/Darknet.py:83]: Remove this function and use self._layer_dict[key] inside __call__ directly.

File: official/vision/beta/projects/yolo/modeling/backbones/Darknet.py [hunk 23]
```diff
@@ -0,0 +1,356 @@
...
+def maxpool_config_todict(config, kwargs):
+  return {
+      "pool_size": config.kernel_size,
+      "strides": config.strides,
+      "padding": config.padding,
+      "name": kwargs["name"]
+  }
+
+
+class layer_registry(object):
```
Comment (arashwan) [inline official/vision/beta/projects/yolo/modeling/backbones/Darknet.py:73]: Can you rename this to layer_factory maybe? Also, can you add a docstr to explain what this does?

File: official/vision/beta/projects/yolo/modeling/backbones/Darknet.py [hunk 24]
```diff
@@ -0,0 +1,356 @@
...
+                            name=f"{config.layer}_{i}")
+        stack_outputs.append(x)
+      elif config.stack == "csp_tiny":
+        x_pass, x = self._tiny_stack(stack_outputs[config.route],
+                                     config,
+                                     name=f"{config.layer}_{i}")
+        stack_outputs.append(x_pass)
+      if (config.is_output and
+          self._min_size == None):
+        endpoints[config.output_name] = x
```
Comment (arashwan) [inline official/vision/beta/projects/yolo/modeling/backbones/Darknet.py:254]: We need output_name to be str, so can you cast it to str everywhere?

File: official/vision/beta/projects/yolo/modeling/backbones/Darknet.py [hunk 25]
```diff
@@ -0,0 +1,356 @@
+import tensorflow as tf
+import tensorflow.keras as ks
+import collections
+
+from official.vision.beta.modeling.backbones import factory
```
Comment (arashwan) [inline official/vision/beta/projects/yolo/modeling/backbones/Darknet.py:5]: We need to add a unit test to make sure the models are created correctly. Can you add a test similar to the ones we have for resnet?

We want to make sure that the number of parameters here match the number of parameters for the official implementation of the paper.

File: official/vision/beta/projects/yolo/modeling/backbones/Darknet.py [hunk 26]
```diff
@@ -0,0 +1,356 @@
...
+    self._bias_regularizer = bias_regularizer
+    self._norm_momentum = norm_momentum
+    self._norm_epislon = norm_epsilon
+    self._use_sync_bn = use_sync_bn
+    self._activation = activation
+    self._weight_decay = kernel_regularizer
+
+    self._default_dict = {
+        "kernel_initializer": self._kernel_initializer,
+        "weight_decay": self._weight_decay,
```
Comment (arashwan) [inline official/vision/beta/projects/yolo/modeling/backbones/Darknet.py:202]: Can you keep the naming kernel_regularizer?

File: official/vision/beta/projects/yolo/modeling/backbones/Darknet.py [hunk 27]
```diff
@@ -0,0 +1,356 @@
...
+    self.bottleneck = bottleneck
+    self.filters = filters
+    self.kernel_size = kernel_size
+    self.strides = strides
+    self.padding = padding
+    self.activation = activation
+    self.route = route
+    self.output_name = output_name
+    self.is_output = is_output
+    return
```
[the remark this answers is not in the document]
Comment (vishnubanna) [reply official/vision/beta/projects/yolo/modeling/backbones/Darknet.py:37]: To clarify would you like us to remove the class?

[the remark this answers is not in the document]
Comment (arashwan) [reply official/vision/beta/projects/yolo/modeling/backbones/Darknet.py:37]: Sorry, no just remove the return line

File: official/vision/beta/projects/yolo/configs/experiments/darknet53.yaml [hunk 28]
```diff
@@ -0,0 +1,53 @@
...
+    input_size: [224, 224, 3]
+    backbone:
+      type: 'darknet'
+      darknet:
+        model_id: 'darknet53'
+  losses:
+    l2_weight_decay: 0.0005
+    one_hot: True
+  train_data:
+    tfds_name: 'imagenet_a'
```
[the remark this answers is not in the document]
Comment (vishnubanna) [reply official/vision/beta/projects/yolo/configs/experiments/darknet53.yaml:17]: that was a copy paste error, for using the same split. when we tested the models, they were different. we will fix that. 

As for the image classification task and TFDS support, we temporarily supported it so that we could test the model because we did not have access to your version of the tfds record. 

Would you like us to add it? We need to change the parser a little bit as well however, because tfds loads the full image, where as your tfrecord decoder loads a string.

File: official/vision/beta/projects/yolo/configs/experiments/darknet53.yaml [hunk 29]
```diff
@@ -0,0 +1,53 @@
+runtime:
+  distribution_strategy: 'mirrored'
+  mixed_precision_dtype: 'float32'
+  loss_scale: 'dynamic'
```
[the remark this answers is not in the document]
Comment (vishnubanna) [reply official/vision/beta/projects/yolo/configs/experiments/darknet53.yaml:4]: for some reason using float16 was stalling on our device, but it was previously tested on float16 and it was working fine. Our image-net dataset down load had a few corrupted files. I will be redownloading it next week to fix the issue.

File: official/vision/beta/projects/yolo/common/registry_imports.py [hunk 30]
```diff
@@ -0,0 +1,23 @@
+# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (vishnubanna) [reply official/vision/beta/projects/yolo/common/registry_imports.py:1]: I think I addressed everything. To run the Darknet test just call 

```
python3 -m official.vision.beta.projects.yolo.modeling.tests.darknet_test
```

File: official/vision/beta/projects/yolo/modeling/backbones/Darknet.py (hunk 26 above)
Comment (vishnubanna) [reply official/vision/beta/projects/yolo/modeling/backbones/Darknet.py]: addressed

File: official/vision/beta/projects/yolo/modeling/backbones/Darknet.py (hunk 25 above)
Comment (vishnubanna) [reply official/vision/beta/projects/yolo/modeling/backbones/Darknet.py:25]: added the test and the doc string, to run the test call

```
python3 -m official.vision.beta.projects.yolo.modeling.tests.darknet_test
```

File: official/vision/beta/projects/yolo/modeling/backbones/Darknet.py (hunk 24 above)
Comment (vishnubanna) [reply official/vision/beta/projects/yolo/modeling/backbones/Darknet.py]: addressed

File: official/vision/beta/projects/yolo/modeling/backbones/Darknet.py (hunk 23 above)
Comment (vishnubanna) [reply official/vision/beta/projects/yolo/modeling/backbones/Darknet.py]: addressed

File: official/vision/beta/projects/yolo/modeling/backbones/Darknet.py (hunk 22 above)
Comment (vishnubanna) [reply official/vision/beta/projects/yolo/modeling/backbones/Darknet.py]: addressed

File: official/vision/beta/projects/yolo/configs/experiments/darknet53.yaml (hunk 28 above)
[the remark this answers is not in the document]
Comment (vishnubanna) [reply official/vision/beta/projects/yolo/configs/experiments/darknet53.yaml]: fixed/addressed

File: official/vision/beta/projects/yolo/configs/experiments/darknet53.yaml (hunk 29 above)
[the remark this answers is not in the document]
Comment (vishnubanna) [reply official/vision/beta/projects/yolo/configs/experiments/darknet53.yaml]: addressed

File: official/vision/beta/configs/backbones.py [hunk 31]
```diff
@@ -85,3 +86,4 @@ class Backbone(hyperparams.OneOfConfig):
   efficientnet: EfficientNet = EfficientNet()
   spinenet: SpineNet = SpineNet()
   mobilenet: MobileNet = MobileNet()
+  darknet: DarkNet = DarkNet()
```
[the remark this answers is not in the document]
Comment (vishnubanna) [reply official/vision/beta/configs/backbones.py]: addressed

File: official/vision/beta/projects/yolo/modeling/building_blocks/_DarkConv.py (hunk 13 above)
Comment (vishnubanna) [reply official/vision/beta/projects/yolo/modeling/building_blocks/_DarkConv.py:35]: I am working on this I did not forget.

File: official/vision/beta/projects/yolo/modeling/tests/test_CSPDownSample.py [hunk 32]
```diff
@@ -0,0 +1,52 @@
+import tensorflow as tf
```
Comment (saberkun) [inline official/vision/beta/projects/yolo/modeling/tests/test_CSPDownSample.py:1]: I think we should put tests together with the library. Please check out the convention of the existing code base.
E.g. modeling/csp_down_sample_test.py

File: official/vision/beta/projects/yolo/modeling/tests/test_CSPConnect.py [hunk 33]
```diff
@@ -0,0 +1,54 @@
+import tensorflow as tf
+import tensorflow.keras as ks
+import numpy as np
+from absl.testing import parameterized
+
+from official.vision.beta.projects.yolo.modeling.building_blocks import CSPDownSample as layer
+from official.vision.beta.projects.yolo.modeling.building_blocks import CSPConnect as layer_companion
```
Comment (saberkun) [inline official/vision/beta/projects/yolo/modeling/tests/test_CSPConnect.py:7]: Please import modules and do not import classes. It is for google py readability.

File: official/vision/beta/projects/yolo/modeling/tests/test_CSPConnect.py [hunk 34]
```diff
@@ -0,0 +1,54 @@
+import tensorflow as tf
```
Comment (saberkun) [inline official/vision/beta/projects/yolo/modeling/tests/test_CSPConnect.py:1]: For all file names, please follow aaa_bbb_ccc.py. No cap. Thanks

Comment (anivegesana) [reply official/vision/beta/projects/yolo/modeling/tests/test_CSPConnect.py:1]: Fixed in 37434ec

File: official/vision/beta/projects/yolo/modeling/tests/test_CSPDownSample.py (hunk 32 above)
Comment (anivegesana) [reply official/vision/beta/projects/yolo/modeling/tests/test_CSPDownSample.py]: Fixed in a0dd0e8

File: official/vision/beta/projects/yolo/modeling/tests/test_CSPConnect.py (hunk 33 above)
Comment (vishnubanna) [reply official/vision/beta/projects/yolo/modeling/tests/test_CSPConnect.py]: fixed in [this commit](https://github.com/tensorflow/models/pull/9405/commits/8ee48095b1e4918478720c688e7703764bffd457)

File: official/vision/beta/projects/yolo/tasks/image_classification.py [hunk 35]
```diff
@@ -0,0 +1,203 @@
...
+from official.core import input_reader
+from official.core import task_factory
+from official.modeling import tf_utils
+from official.vision.beta.projects.yolo.configs import darknet_classification as exp_cfg
+from official.vision.beta.dataloaders import classification_input
+from official.vision.beta.modeling import factory
+
+
+@task_factory.register_task_cls(exp_cfg.ImageClassificationTask)
+class ImageClassificationTask(base_task.Task):
```
[the remark this answers is not in the document]
Comment (vishnubanna) [reply official/vision/beta/projects/yolo/tasks/image_classification.py]: addressed in 47545935c167c28f66896768481e7d25a870eaff

File: official/vision/beta/projects/yolo/modeling/layers/csp_connect_test.py [hunk 36]
```diff
@@ -0,0 +1,52 @@
+import tensorflow as tf
+import tensorflow.keras as ks
```
[the remark this answers is not in the document]
Comment (vishnubanna) [reply official/vision/beta/projects/yolo/modeling/layers/csp_connect_test.py]: addressed in 47545935c167c28f66896768481e7d25a870eaff

File: official/vision/beta/projects/yolo/modeling/layers/__init__.py [hunk 37]
```diff
@@ -0,0 +1 @@
+from .nn_blocks import Identity, CSPTiny, CSPDownSample, CSPConnect, DarkTiny, DarkResidual, DarkConv
```
[the remark this answers is not in the document]
Comment (vishnubanna) [reply official/vision/beta/projects/yolo/modeling/layers/__init__.py]: addressed in 47545935c167c28f66896768481e7d25a870eaff

**Review by arashwan — COMMENTED** [review]
remove coco/instances_val2017.json from the commit.

File: coco_retinanet_gpu.yaml [hunk 38]
```diff
@@ -0,0 +1,71 @@
...
+  validation_interval: 2000
+  steps_per_loop: 200 #59136
+  summary_interval: 200 #59136
+  checkpoint_interval: 10000
+  optimizer_config:
+    optimizer:
+      type: 'sgd'
+      sgd:
+        momentum: 0.9
+        #    learning_rate:
```
Comment (arashwan) [inline coco_retinanet_gpu.yaml:53]: Can you remove the commented learning rate?

File: nohup.out [hunk 39]
```diff
@@ -0,0 +1,987 @@
+2020-10-30 19:06:27.235259: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudart.so.10.1
```
Comment (arashwan) [inline nohup.out:1]: remove this file.

File: official/modeling/activations/__init__.py [hunk 40]
```diff
@@ -16,6 +16,7 @@
 from official.modeling.activations.gelu import gelu
 from official.modeling.activations.relu import relu6
 from official.modeling.activations.sigmoid import hard_sigmoid
+
```
Comment (arashwan) [inline official/modeling/activations/__init__.py:19]: Leave this file unchanged as we discussed.

File: official/vision/beta/projects/yolo/common/registry_imports.py [hunk 41]
```diff
@@ -0,0 +1,28 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""All necessary imports for registration."""
+
+# pylint: disable=unused-import
+from official.nlp import tasks as nlp_task
```
Comment (arashwan) [inline official/vision/beta/projects/yolo/common/registry_imports.py:18]: Just import registry_imports here and everything inside it will be imported.
i.e. from official.... import registery_imports
Then you can add your additional imports after.

File: official/vision/beta/train.py [hunk 42]
```diff
@@ -35,6 +35,8 @@
 def main(_):
   gin.parse_config_files_and_bindings(FLAGS.gin_file, FLAGS.gin_params)
   params = train_utils.parse_configuration(FLAGS)
+  import pprint
+  pprint.pprint(params.as_dict())
```
Comment (arashwan) [inline official/vision/beta/train.py:39]: leave this file unchanged.

**Review by arashwan — COMMENTED** [review]
Can you remove training_dir files as well?

File: official/vision/beta/projects/yolo/modeling/layers/nn_blocks.py [hunk 43]
```diff
@@ -0,0 +1,768 @@
...
+      **kwargs: Keyword Arguments
+  '''
+    def __init__(
+            self,
+            filters=1,
+            kernel_size=(1, 1),
+            strides=(1, 1),
+            padding='same',
+            dilation_rate=(1, 1),
+            use_bias=True,
```
Comment (arashwan) [inline official/vision/beta/projects/yolo/modeling/layers/nn_blocks.py:54]: It looks like use_bias is determined by the use of batch norm, so we can remove this from the list of arguments, right?

File: official/vision/beta/projects/yolo/modeling/layers/nn_blocks.py [hunk 44]
```diff
@@ -0,0 +1,768 @@
...
+      sc_activation: string for activation function to use in layer
+      downsample: boolean for if image input is larger than layer output, set downsample to True
+                  so the dimensions are forced to match
+      **kwargs: Keyword Arguments
+
+  '''
+    def __init__(self,
+                 filters=1,
+                 filter_scale=2,
+                 use_bias=True,
```
Comment (arashwan) [inline official/vision/beta/projects/yolo/modeling/layers/nn_blocks.py:208]: Again if use_bn is True, are you still using bias?

File: official/vision/beta/projects/yolo/README.md [hunk 45]
```diff
@@ -0,0 +1,52 @@
+# YOLO Object Detectors, You Only Look Once
+
```
Comment (arashwan) [inline official/vision/beta/projects/yolo/README.md:2]: You need to limit the number of characters per line to 80 whenever possible.

**Review by arashwan — APPROVED** [review]
Thanks guys for the hard work!

**Review by jaeyounkim — APPROVED** [review]
